File: JavaNgArch/src/javangarch/mvc/home/HomeController.java
package javangarch.mvc.home;

import java.util.ArrayList;
import java.util.Date;
import java.util.Map;
import java.util.stream.Collectors;

import javax.annotation.Resource;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import javax.validation.Valid;

import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import projectbase.bd.BizException;
import projectbase.domain.ICommonBD;
import projectbase.domain.customcollection.IDomainList;
import projectbase.mvc.BaseController;
import projectbase.mvc.DisplayExtension;
import projectbase.mvc.Transaction;
import projectbase.mvc.WebConfigurationManager;
import projectbase.mvc.result.ActionResult;
import projectbase.utils.Util;
import javangarch.domain.bdinterface.IAdminBD;
import javangarch.domain.domainmodel.gn.Corp;
import javangarch.domain.domainmodel.gn.Dept;
import javangarch.domain.domainmodel.gn.Func;
import javangarch.domain.domainmodel.gn.User;
import javangarch.mvc.shared.AppBaseController;

public class HomeController extends BaseController {
	@Resource
	public IAdminBD AdminBD;

	@Resource
	public ICommonBD<Corp> CorpBD;
	@Resource
	public ICommonBD<Dept> DeptBD;
	@Resource
	public ICommonBD<User> UserBD;

	
	public ActionResult ShowLogin(HttpSession session) {
		if (session.getAttribute("LoginInfo") != null)
			return ClientRedirect("MainFrameLoggedIn");
		LoginAttemptViewModel m=new LoginAttemptViewModel();
		m.setCorpCode("1");
		m.setCode("test");
		m.setPassword("1");
		return ForView("Login", NoMaster,m );
	}

	@RequestMapping(method = RequestMethod.POST)
	@Transaction
	public ActionResult Login(@Valid LoginAttemptViewModel attempt,
			HttpSession session,
			@CookieValue(required = false) Cookie loginMark,
			HttpServletResponse response) {
		if (session.getAttribute("LoginInfo") != null)
			return ClientRedirect("MainFrameLoggedIn");

		User loginUser = AdminBD.GetLoginUser(attempt.getCorpCode(),
				attempt.getCode(), attempt.getPassword());
		if (loginUser == null) {
			SetViewMessage("Login_Wrong");
			return ClientShowMessage();
		}
		if (!loginUser.getIsActive()) {
			SetViewMessage("Login_IsNotActive");
			return ClientShowMessage();
		}

	/*	if (loginUser.getLoginMark() != null) {// 检查本次登录是否为上次的登录的会话延续
			if (loginMark == null)
				return ClientShowMessage("Login_AlreadyIn");
			String lastsessionId = loginMark.getValue();
			if (!lastsessionId.equalsIgnoreCase(loginUser.getLoginMark()))
				return ClientShowMessage("Login_AlreadyIn");
		}

		loginMark = new Cookie("LoginMark", session.getId());
		loginMark.setHttpOnly(true);
		loginMark.setMaxAge(2 * 24 * 60 * 60);
		response.addCookie(loginMark);

		loginUser.setLoginMark(session.getId());*/
		LoginInfoViewModel loginInfo = new LoginInfoViewModel();
		loginInfo.setLoginUser(loginUser);
		loginInfo.setLoginCorp(loginUser.getDept().getCorp());
		String tmp = loginInfo.getLoginCorp().getName();// make proxy load the
														// real entity
		int i = 0;
		for (Func func : loginInfo.getLoginCorp().getFuncs()) {
			loginInfo.AddCorpFuncCode(func.getCode());
		}
		for (Func func : loginUser.getDept().getFuncs()) {
			loginInfo.AddDeptFuncCode(func.getCode());
		}
		for (Func func : loginUser.getFuncs()) {
			loginInfo.AddUserFuncCode(func.getCode());
		}
		session.setAttribute("LoginInfo", loginInfo);

		Util.AddLog("login=" + loginInfo.getLoginUser().getId());
		return ClientRedirect("MainFrameLoggedIn");
	}

	public ActionResult MainFrame(HttpSession session) {
		LoginInfoViewModel loginInfo = ((LoginInfoViewModel) session
				.getAttribute("LoginInfo"));
		if (loginInfo != null){
			AdminBD.RefreshUser(loginInfo.getLoginUser());
		}
		return  ForView(null, NoMaster, loginInfo);
	}

	public ActionResult Logout(HttpSession session) {
		session.invalidate();
		return ClientRedirect("ShowLogin");
	}


}

File: NetNgArch/src/NetNgArch.Web.Mvc/Home/HomeController.cs
using System.Collections;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using NetNgArch.Domain.BDInterface;
using NetNgArch.Domain.DomainModel.GN;
using ProjectBase.Domain;
using ProjectBase.Utils;
using ProjectBase.Web.Mvc;
using System.Collections.Generic;
using System;
using ProjectBase.Web.Mvc.Angular;


namespace NetNgArch.Web.Mvc.Home
{
    public class HomeController : BaseController
    {
        public IAdminBD AdminBD { get; set; }
        public ICommonBD<Corp> CorpBD { get; set; }
        public ICommonBD<Dept> DeptBD { get; set; }
        public ICommonBD<User> UserBD { get; set; }

        public ActionResult ShowLogin()
        {
            if (Session["LoginInfo"] != null) return ClientRedirect("MainFrameLoggedIn");
            var m = new LoginAttemptViewModel()
                        {
                            CorpCode = "1",
                            Code = "admin",
                            Password = "1"
                        };
            return ForView("Login", m);
        }
        [HttpPost]
        [Transaction]
        public ActionResult Login(LoginAttemptViewModel attempt)
        {
            if (Session["LoginInfo"] != null) return ClientRedirect("MainFrameLoggedIn");

            var loginUser = AdminBD.GetLoginUser(attempt.Code, attempt.Password);
            if (loginUser == null)
            {
                SetViewMessage("Login_Wrong");
                return ClientShowMessage();
            }
            if (!loginUser.IsActive)
            {
                SetViewMessage("Login_IsNotActive");
                return ClientShowMessage();
            }
            var loginMark = Request.Cookies["LoginMark"];
            if (loginUser.LoginMark!=null){//检查本次登录是否为上次的登录的会话延续
                if (loginMark==null) return ClientShowMessage("Login_AlreadyIn");
                var lastsessionId = loginMark["lastsessionId"];
                if (lastsessionId != loginUser.LoginMark) return ClientShowMessage("Login_AlreadyIn");
            }

            loginMark = new HttpCookie("LoginMark");
            loginMark.HttpOnly = true;
            loginMark["lastsessionId"] = Session.SessionID;
            loginMark.Expires = DateTime.Now.AddDays(2);
            Response.Cookies.Add(loginMark);
           
            loginUser.LoginMark = Session.SessionID;
            var loginInfo = new LoginInfoViewModel();
            loginInfo.LoginUser = loginUser;
            loginInfo.LoginCorp = loginUser.Dept.Corp;
            var tmp = loginInfo.LoginCorp.Name;//make proxy load the real entity
            var i = 0;
            foreach(var func in loginInfo.LoginCorp.Funcs)
            {
                loginInfo.AddCorpFuncCode(func.Code);
            }
            foreach (var func in loginUser.Dept.Funcs)
            {
                loginInfo.AddDeptFuncCode(func.Code);
            }
            foreach (var func in loginUser.Funcs)
            {
                loginInfo.AddUserFuncCode(func.Code);
            }
            Session["LoginInfo"] = loginInfo;
            return ClientRedirect("MainFrameLoggedIn");
        }
    
        public ActionResult MainFrame()
        {
            Util.test();
            var loginInfo = ((LoginInfoViewModel) Session["LoginInfo"]);
            if (loginInfo != null)
            {
                AdminBD.RefreshUser(loginInfo.LoginUser);
            }
            return ForView(loginInfo);
        }
        [HttpPost]
        public ActionResult Logout()
        {
            Session.Abandon();
            return ClientReloadApp("MainFrame",null);
        }
    }
}
